Docs: Split Subprojects in Explanation and How-to (Diátaxis) (#9785)
readthedocs/readthedocs.org: docs/user/subprojects.rst

> Organizing projects in one site: Subprojects

Author: I think it's okay to capitalize after a `:`? :)

I was also thinking about saying "multiple" or "many" here but I wanted to trim it because of length. But now that you mention it, I feel confident to add it :+1:

> Organizing projects in one site: Subprojects

Author: (edited your suggestion to capital Subprojects)

> Organizing multiple projects in one site: Subprojects
> ============================================

Author: ```suggestion
Organizing multiple projects in one site: Subprojects
=====================================================
```

Projects can be configured in a nested manner, by configuring a project as a
*subproject* of another project. This allows for documentation projects to share
a search index and a namespace or custom domain, but still be maintained
independently.

This is useful for:

* Organizations that need all their projects visible in one documentation portal or landing page
* Projects that document and release several packages through a combined release cycle

> * Projects that document and release several packages through different release cycles

Author: Let's say we have:

* Main Application documentation
  * Extension Foo documentation (subproject)
  * Extension Bar documentation (subproject)

Then we could have:

a) a Main Application that ships with official extensions bundled in the same release
b) a Main Application and all extensions aligning their version numbers by convention (rare case)
c) separate release cycles but common documentation site for both Main Application and extensions (probably the common case)

Actually the two items are spelling out that whatever the release cycles are for these 3 projects, this is still supported?

AFAICT, Nautobot is a good demo of c). If you visit https://docs.nautobot.com/projects/golden-config/en/latest/, you will see that the application has a separate version drop down from the main project.

Do we have experience with this? I'm working with [a large project on GitBook](https://govstack.gitbook.io/specification/building-blocks/about-building-blocks) and the version dropdown on GitBook changes, depending on whether the user is on the main project or the subproject (what GitBook calls a collection).

> * Projects that document and release several packages through different release cycles

Reviewer: I don't think that "combined" and "different" release cycles communicate in a clear way what you just explained to me (thanks for the explanation 💯 , btw). I don't know if they are common terms, but I'm not aware of them and that is why I was pretty confused originally. If we don't find a simpler way to communicate this, I'd remove these bullets.

> Do we have experience with this? I'm working with [a large project on GitBook](https://govstack.gitbook.io/specification/building-blocks/about-building-blocks) and the version dropdown on GitBook changes, depending on whether the user is on the main project or the subproject (what GitBook calls a collection).

IIRC, the flyout menu belongs to the project itself. So, if you visit the main project, you will get the version of the main project. If you visit the subproject, you will get the version for the subproject.

> * Projects that document and release several packages through different release cycles

Author: @humitos thanks! I think your experience aligns with what I see on Nautobot, so I'm gonna go ahead and write something that mentions the flexibility... but also in a simpler way. It's really nice to know that the flyout menu has this functionality.

> * Projects that document and release several packages through different release cycles

Author: I think I have solved this by writing a separate subsection that only focuses on "Separate release cycles". I am quite sure this is a) quite normal and b) a complexity that people worry about when they think about combining multiple documentation sets.

The above list has been cleaned up and there's a subsection further down :+1:

> Organizing multiple projects in one site: Subprojects

Reviewer 2: I'd put the feature name at the front:

```suggestion
Subprojects: host multiple projects on a single domain
```

Author: I thought it was so simple and that all words following a colon would be capitalized -- wow I was wrong :upside_down_face: https://www.grammarly.com/blog/capitalization-after-colons/

(also this is correct)

=====================================================

> Projects can be configured in a nested manner, by configuring a project as a

Reviewer 2: `nested manner` and `namespace` is not user-friendly language, we should improve this somehow.

Author: Yup, this was also missing the "In this article, you can learn about..." softer introduction. I added this, so now there's yet another abstraction layer.

*subproject* of another project. This allows for documentation projects to share
a search index and a namespace or custom domain, but still be maintained
independently.

This is useful for:

* Organizations that need all their projects visible in one documentation portal or landing page
* Projects that document and release several packages or extensions
* Organizations or projects that want to have a common search function for several sets of documentation

For example, a parent project, ``foo`` is set up with a subproject, ``bar``. The

> documentation for ``foo`` will be available at:

Reviewer 2: I'd love to see a more real example here. Perhaps `example-project` and `example-api` or something?

Author: I used `example-project` and `plugin` :)

If the example project ``example-project`` was set up with a custom domain,
``docs.example.com``, the URLs for projects ``example-project`` and ``example-project-plugin`` with alias ``plugin`` would
respectively be at:

* ``example-project``: https://docs.example.com/en/latest/
* ``example-project-plugin``: https://docs.example.com/projects/plugin/en/latest/

Using aliases
-------------

Adding an alias for the subproject allows you to override the URL that is used to access it,
giving more control over how you want to structure your projects.
You can choose an alias for the subproject when it is created.

You can set your subproject's project name and :term:`slug` however you want,
but we suggest prefixing it with the name of the main project.

Typically, a subproject is created with a ``$mainproject-`` prefix,
for instance if the main project is called ``example-project`` and the subproject is called ``plugin``,
then the subproject's Read the Docs project :term:`slug` will be ``example-project-plugin``.
When adding the subproject,
the alias is set to ``plugin`` and the project's URL becomes
``example-project.readthedocs.io/projects/plugin``.

> When you add a subproject,
> the subproject will not be directly available anymore from its own domain.

Author: ```suggestion
When a project becomes a subproject,
the subproject, its original domain will automatically redirect to the subproject's URL.
```

Author: I have a hard time writing this message in a very clear way, but I think that the example right after it makes it okay.

Any quick suggestions @humitos ? Otherwise, I thought to merge this, as its the last outstanding item :)

Reviewer: Much better, yeah!

Subprojects: host multiple projects on a single domain
======================================================

In this article, you can learn more about how several documentation projects can be combined and presented to the reader on the same website.

> Read the Docs projects can be configured to make other *subprojects* available on the website of the *main project*.

Reviewer: ```suggestion
Read the Docs can be configured to make other *projects* available on the website of the *main project* as **subprojects**.
```

Author: Really good tweak :+1:

This allows for documentation projects to share a search index and a namespace or custom domain,
but still be maintained independently.

This is useful for:

* Organizations that need all their projects visible in one documentation portal or landing page
* Projects that document and release several packages or extensions
* Organizations or projects that want to have a common search function for several sets of documentation

> For a main project ``example-project``, a subproject ``example-project-plugin`` can be made available as follows:

Reviewer: This sentence is hard to read for me. I'm not sure if "can be made available" is what confuses me, there is a word missing or something else 🙃  --maybe I'm just tired.

Author: Let's fix it -- being tired is a good benchmark for readability :+1:

Reviewer 2: Yea, `can be made available` is super passive voice. `Is activated` or similar would be active voice.

* Main project: https://example-project.readthedocs.io/en/latest/
* Subproject: https://example-project.readthedocs.io/projects/plugin/en/latest/

.. seealso::

   :doc:`/guides/subprojects`
     Learn how to create and manage subprojects

   :doc:`/guides/intersphinx`

>      Learn how to use references between different Sphinx projects, for instance between subprojects

Reviewer: I really like how these `seealso` render and work.

Sharing a custom domain
-----------------------

Projects and subprojects can be used to share a custom domain.

> To configure this, one project should be established as the main project.
> This project will be configured with a custom domain.

Reviewer: ```suggestion
To configure this, one project should be established as the main project and configured with a custom domain.
```

Other projects are then added as subprojects to the main project.

If the example project ``example-project`` was set up with a custom domain,
``docs.example.com``, the URLs for projects ``example-project`` and ``example-project-plugin`` with alias ``plugin`` would
respectively be at:

* ``example-project``: https://docs.example.com/en/latest/
* ``example-project-plugin``: https://docs.example.com/projects/plugin/en/latest/

Using aliases
-------------

> You can choose an alias for the subproject when it is created.
> This allows you to override the URL that is used to access it,
> giving more control over how you want to structure your projects.

Reviewer: I'd invert this to mention the benefit first.

```suggestion
Adding an alias for the subproject allows you to override the URL that is used to access it,
giving more control over how you want to structure your projects.
You can choose an alias for the subproject when it is created.
```

Reviewer 2: The marketing discussion is seeping in :D

Reviewer: 🧑‍🎓

You can set your subproject's project name and :term:`slug` however you want,
but we suggest prefixing it with the name of the main project.

> Typically, a subproject is created with a ``$mainproject-`` prefix,

Reviewer: ```suggestion
Typically, a subproject is created with a ``mainproject-`` prefix,
```

Is the `$` on purpose here?

Author: Yes, well, I am a bit clueless about how to write placeholders.

@ericholscher do you have a favorite? `<placeholder>` or `$placeholder` or `PLACEHOLDER` or some other variant?

Author: (would like to immediately add it to the style guide as well - this keeps popping up)

Reviewer 2: I think `<Thing>` for something abstract (eg. `<Username dropdown>`) and `$mainproject` for a string substitution would be my initial thinking.

Reviewer: I think we have been using `<thing>` in most of the pages in the documentation. The only place were we use `$` are redirects and outgoing webhooks --which is a more technical audience that may be more used to what `$` could mean.

That doesn't mean that `<thing>` is a lot clearer for non-technical audiences, but at least keeps consistency with other pages: https://docs.readthedocs.io/en/stable/hosting.html#subdomain-support and https://docs.readthedocs.io/en/latest/custom-domains.html

That said, I think we need a `<thing>` entry in our glossary explaining that `<>` means a substitution done by the user/reader 👍🏼

Author: I tend to agree with Manuel that maybe using `<thing>` more or less for everything is simpler but would like to see it as a list of examples.

For this article, it will make it consistent with the rest of the article... so we can _somewhat easily_ find these occurrences and update them after we've finished a decision in the Documentation Style Guide.

```suggestion
Typically, a subproject is created with a ``<mainproject>-`` prefix,
```

Author: Just flagging that I started typing in a style guide update, BRB...

> for instance if the main project is called ``example-project-plugin`` and the subproject is called ``plugin``,

Reviewer: ```suggestion
for instance if the main project is called ``example-project`` and the subproject is called ``plugin``,
```

> then the subproject's Read the Docs project name will be ``example-project-plugin``.

Reviewer: ```suggestion
then the subproject's Read the Docs project slug will be ``example-project-plugin``.
```

Author: Updated with `:term:` :+1:

When adding the subproject,
the alias is set to ``plugin`` and the project's URL becomes
``example-project.readthedocs.io/projects/plugin``.

> When you add a subproject,

Reviewer: ```suggestion
When you add an alias to a subproject,
```

Author: Isn't the following part true regardless of adding an alias?

Author: Or do you mean because of the `/projects/plugin`?

I think we want to tell the reader mainly that the original domain `example-project-plugin.readthedocs.io` now redirects (regardless of the alias)

Reviewer: Got it! In that case, I think we should merge this paragraph with the previous one since both talk about the same action "adding a subproject" and "where the project is gonna be hosted" now.

Author: Fused the two suggestions into a simpler one.

> the subproject will not be directly available anymore from its own domain.
> For instance, ``example-project-plugin.readthedocs.io/`` will redirect to ``example-project.readthedocs.io/projects/plugin``.

Reviewer: It will still be available from it owns domain, but it will redirect.

```suggestion
Read the Docs will redirect the old subproject URL to the new one using the main project domain and the alias in the path.
For instance, ``example-project-plugin.readthedocs.io/`` will redirect to ``example-project.readthedocs.io/projects/plugin``.
```

Author: I get it now, saying "not be directly available" contradicts the redirect, since a redirect strictly speaking makes it available :+1:

> For instance, ``example-project-plugin.readthedocs.io/`` will redirect to ``example-project.readthedocs.io/projects/plugin``.

Author: ```suggestion
For instance, ``example-project-plugin.readthedocs.io/`` will redirect to ``example-project.readthedocs.io/projects/<alias or subproject slug>``.
```

Subprojects: host multiple projects on a single domain
======================================================

In this article, you can learn more about how several documentation projects can be combined and presented to the reader on the same website.

Read the Docs projects can be configured to make other *subprojects* available on the website of the *main project*.
This allows for documentation projects to share a search index and a namespace or custom domain,
but still be maintained independently.

This is useful for:

* Organizations that need all their projects visible in one documentation portal or landing page
* Projects that document and release several packages or extensions
* Organizations or projects that want to have a common search function for several sets of documentation

> For example, a parent project, ``example-project`` is set up with a subproject, ``plugin``. The

Reviewer 2: Should we follow our own advice and name it `example-project-plugin`, with an alias of `plugin`?

documentation for ``example-project`` will be available at:

https://example-project.readthedocs.io/en/latest/

The documentation for ``plugin`` will be available under this same path:

https://example-project.readthedocs.io/projects/plugin/en/latest/

.. seealso::

   :doc:`/guides/subprojects`

Sharing a custom domain
-----------------------

Projects and subprojects can be used to share a custom domain.
To configure this, one project should be established as the main project.
This project will be configured with a custom domain.

> Other projects are then added as subprojects to the main project.

Reviewer 2: This is great. Shows the main benefit, and how to achieve it very quickly.

Author: Yes, and as a bonus, it says "subproject" and "main project" quite a lot :)

If the example project ``foo`` was set up with a custom domain,
``docs.example.com``, the URLs for projects ``foo`` and ``bar`` would
respectively be at:

* `foo`: https://docs.example.com/en/latest/
* `bar`: https://docs.example.com/projects/bar/en/latest/

.. note::

   The terms "parent" and "child" are also used,
   where the main project is the parent and all subprojects are a child.
   You can **not** have a child of a child,

>    meaning that you cannot add a subproject to a subproject.

Reviewer 2: This is great, but we should probably stick to a single language if possible. 👍

Author: The reason for the note was that we actually have "child" in the form field.

Should I cross a line and rename the form label directly in this PR? :see_no_evil:

Author: Ah, okay, no, but I can open a separate PR and remove the whole note admonition

Using aliases
-------------

You can choose an alias for the subproject when it is created.
This allows you to override the URL that is used to access it,
giving more control over how you want to structure your projects.

You can set your subproject's project name and :term:`slug` however you want,
but we suggest prefixing it with the name of the main project.

Typically, a subproject is created with a ``<mainproject>-`` prefix,
for instance if the main project is called ``foo`` and the subproject is called ``bar``,
then the subproject's Read the Docs project name will be ``foo-bar``.
When adding the subproject,
the alias is set to ``bar`` and the project's URL becomes

> ``foo.readthedocs.io/projects/bar``.

Reviewer 2: This is a really clear explanation 👍 I do wonder about using real examples here too, instead of foo/bar, but it's less important I think.

Author: The use of example project names is consistent now. The foos and the bars are gone.

When you add a subproject,
the subproject will not be directly available anymore from its own domain.

> For instance, ``subproject.readthedocs.io/`` will redirect to ``mainproject.readthedocs.io/projects/subproject``.

Reviewer 2: 👍

project. This project will be configured with a custom domain. Projects can then
be added as subprojects to this parent project.

If the example project ``foo`` was set up with a custom domain,
``docs.example.com``, the URLs for projects ``foo`` and ``bar`` would

> respectively be at: https://docs.example.com/en/latest/ and
> https://docs.example.com/projects/bar/en/latest/

Reviewer 2: ```suggestion
respectively be at: 

* `foo`: https://docs.example.com/en/latest/
* `bar`: https://docs.example.com/projects/bar/en/latest/
```

since your documentation will always be served from
the domain of the parent project.

Separate release cycles
-----------------------

> When several projects are documented in the same space, it's common that they follow separate release cycles.

Reviewer 2: I don't know what this means? What is a `space`? How does it relate to the release cycle?

Author: I removed the term `space` which I guess is something I have been indoctrinated to say casually from academia :D

We recommend that the documentation follows the release cycle of whatever it is documenting.
By using subprojects, you can present your documentation for several projects with separate release cycles.
This is solved by having the :term:`flyout menu` active for the project that's viewed.

When the user navigates to a subproject,

> they are presented with a :term:`flyout menu` matching the subproject's versions and :doc:`/downloadable-documentation`.

Reviewer 2: Weird to use the `term` twice in such quick succession I think.

Overall I'm not quite sure what this whole section is trying to say? I think an example or screenshot could help, but I don't get the value it's trying to show? That there's a flyout, and it exists?

Author: I've updated it now. Does the point stand out clearer? I think the point is very critical to make because it's definitely one of the reasons that I think this feature is very useful (and because I think _not_ supporting separate release cycles would render it fairly useless).

Separate release cycles
-----------------------

By using subprojects,
you can present the documentation of several projects
even though they have separate release cycles.

Your main project may have its own versions and releases,
while all of its subprojects maintain their own individual versions and releases.
We recommend that documentation follows the release cycle of whatever it is documenting,

> meaning that your subprojects should be free to follow their own release cycle.

Reviewer 2: This is much clearer 💯
